src/devices/sound/discrete.cpp

```diff
@@ -1017,22 +1025,18 @@ void discrete_device::process(int samples)
 	if (samples == 0)
 		return;
 
-	/* Setup tasks */
+	// Set up tasks
 	for (const auto &task : task_list)
-	{
-		/* unlock the thread */
-		task->unlock();
-
 		task->prepare_for_queue(samples);
-	}
+	std::atomic_thread_fence(std::memory_order_release);
```

Comment:
The unlock operation has been moved inside `discrete_task::prepare_for_queue`.  The release barrier here ensures preparation work can’t be reordered after starting the tasks.